lib/firenest/replicated_state/remote.ex

```diff
@@ -0,0 +1,219 @@
+
+  # TODO: some protocol for requesting more of a state, even from other nodes
+  # on first up, so we don't need to go to each node separately.
+
+  import Record
+  defrecord :deltas, max: nil, lowest: 0, store: %{}
+
+  def new(:ignore, broadcast, max_deltas)
+      when is_function(broadcast, 0) and is_integer(max_deltas) do
+    %__MODULE__{tag: :ignore, broadcast: broadcast, deltas: deltas(max: max_deltas)}
+  end
+
+  def clock(%__MODULE__{clock: clock}), do: clock
+
+  # Reconnections are dead until we have permdown
+  def up(%__MODULE__{clocks: clocks} = state, ref, clock) do
+    case clocks do
+      # Reconnection, try to catch up
+      %{^ref => old_clock} when clock > old_clock ->
+        {:catch_up, old_clock, state}
+
+      # Reconnection, no remote changes
+      %{^ref => old_clock} ->
+        # Assert for sanity
+        true = old_clock == clock
+        {:ok, state}
+
+      # New node, no state
+      %{} when clock == 0 ->
+        {:ok, %{state | clocks: Map.put(clocks, ref, clock)}}
+
+      # New node, catch up
+      %{} ->
+        {:catch_up, 0, state}
+    end
+  end
+
+  # TODO: Right now down means permdown
+  def down(state, ref) do
+    permdown(state, ref)
+  end
+
+  def permdown(%__MODULE__{clocks: clocks} = state, ref) do
+    true = Map.has_key?(clocks, ref)
+    clocks = Map.delete(clocks, ref)
+    {:delete, ref, %{state | clocks: clocks}}
+  end
+
+  def catch_up(%__MODULE__{clock: current} = state, old_clock, state_getter)
+      when old_clock < current do
+    %{deltas: deltas, tag: tag} = state
+
+    if Map.has_key?(deltas, old_clock) do
+      {:deltas, tag, current, Enum.flat_map(old_clock..current, &Map.fetch!(deltas, &1))}
+    else
+      {:state_transfer, tag, current, state_getter.()}
+    end
+  end
+
+  def broadcast(%__MODULE__{} = state, prepare_delta) do
+    %{pending: pending, clock: clock, tag: tag, broadcast: broadcast, deltas: deltas} = state
+    {:scheduled, broadcast} = broadcast
+
+    new_deltas = prepare_deltas(tag, pending, prepare_delta)
+
+    new_clock = clock + 1
+    deltas = store_deltas(deltas, new_clock, new_deltas)
+    new_state = %{state | pending: %{}, clock: new_clock, broadcast: broadcast, deltas: deltas}
+
+    {{tag, clock, new_deltas}, new_state}
+  end
+
+  def handle_catch_up(%__MODULE__{tag: tag} = state, ref, {:deltas, tag, clock, deltas}) do
+    %{clocks: clocks} = state
+
+    state = %{state | clocks: %{clocks | ref => clock}}
+    {puts, updates, deletes} = handle_deltas(tag, deltas)
+    {:diff, puts, updates, deletes, state}
+  end
+
+  def handle_catch_up(%__MODULE__{tag: tag} = state, from, {:state_transfer, tag, clock, data}) do
+    %{clocks: clocks} = state
+
+    case tag do
+      :ignore -> {:insert, data, %{state | clocks: %{clocks | from => clock}}}
+    end
+  end
+
+  # TODO: should we store somewhere we're catching up with the server?
+  # if so, then we should accumulate the broadcasts we can't handle, until we can.
```

Comment:
Yes, we always needs to this because I believe the topology does not guarantee ordering between sends and broadcasts (or it should not, as that by definition implies using the same connection). So if you need ask a node to catch you up and it replies to you and then broadcasts, there is a chance the broadcast will arrive before the reply, and you would have lost it.